File: 卡码/代码随想录/Python/栈与队列/stackAndQueue/04-滑动窗口/LeetCode_0239.py
#239. 滑动窗口最大值
from typing import *
from collections import deque

class Queue:
  def __init__(self):
    self.queue=deque()
  def front(self)->int:
    return self.queue[0]
  def back(self)->int:
    return self.queue[-1]
  def empty(self):
    return len(self.queue)==0
  def push(self,val):
    while self.queue and val>self.back():
      self.queue.pop()
    self.queue.append(val)
  def pop(self,val):
    if self.queue and val==self.front():
      self.queue.popleft()

class Solution:
  def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
    res=[]
    queue=Queue()
    for i in range(k):
      queue.push(nums[i])
    res.append(queue.front())
    for i in range(k,len(nums)):
      queue.pop(nums[i-k])
      queue.push(nums[i])
      res.append(queue.front())
    return res

File: 卡码/代码随想录/1/05-栈与队列/Go/04-滑动窗口/LeetCode_0239.go
//239. 滑动窗口最大值
package main

import "fmt"

type Queue struct{
	queue []int
}
func (this *Queue) front() int{
	return this.queue[0]
}
func (this *Queue) back() int{
	return this.queue[len(this.queue)-1]
}
func (this *Queue) empty() bool{
	return len(this.queue)==0
}
func (this *Queue) push(val int){
	for !this.empty()&&val>this.back() {
		this.queue=this.queue[:len(this.queue)-1]
	}
	this.queue=append(this.queue,val)
}
func (this *Queue) pop(val int){
	if !this.empty()&&this.front()==val {
		this.queue=this.queue[1:]
	}
}

func maxSlidingWindow(nums []int, k int) []int {
	queue:=&Queue{make([]int,0)}
	res:=make([]int,0)
	for i:=0;i<k;i++ {
		queue.push(nums[i])
	}
	res=append(res,queue.front())
	for i:=k;i<len(nums);i++ {
		queue.pop(nums[i-k])
		queue.push(nums[i])
		res=append(res,queue.front())
	}
	return res
}

func main(){
	fmt.Println(maxSlidingWindow([]int{1,3,-1,-3,5,3,6,7},3))
}
